File: 3. Longest String without Repeating Characters/lengthOfLongestSubstring.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        char[] sChar = s.toCharArray();
        HashMap<Character, Integer> map = new HashMap<>();
        int left = 0;
        int result = 0;

        for (int i = 0; i < sChar.length; i++) {
            if (map.containsKey(sChar[i]) && map.get(sChar[i]) >= left) {
                int j = map.get(sChar[i]);
                left = j + 1;
            }
            map.put(sChar[i], i);
            result = Math.max(result, i - left + 1);
        }
        return result;
    }
}

File: 3. Longest String without Repeating Characters/lengthOfLongestSubstring.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map = {}
        left = 0
        right = 0
        result = 0
        while right < len(s):
            if s[right] in map and map[s[right]] >= left:
                left = map[s[right]] + 1
            map[s[right]] = right
            result = max(result, right - left + 1)
            right += 1
        return result